File: bookstore-py/services/BookStoreService/v1/BookStoreService.py
"""sylk.build service implemantation for -> BookStoreService"""
import grpc
from google.protobuf.timestamp_pb2 import Timestamp
from typing import Iterator
from google.protobuf import empty_pb2
from services.protos.public.bookstore.book.v1 import book_pb2_grpc, book_pb2

class BookStoreService(book_pb2_grpc.BookStoreServiceServicer):

	# @rpc @@sylk - DO NOT REMOVE
	def ListBooks(self, request: empty_pb2.Empty, context: grpc.ServicerContext) -> Iterator[book_pb2.Book]:
		# responses = [book_pb2.Book(name=None)]
		# for res in responses:
		#    yield res

		super().ListBooks(request, context)

	# @rpc @@sylk - DO NOT REMOVE
	def GetBook(self, request: book_pb2.GetBookRequest, context: grpc.ServicerContext) -> book_pb2.Book:
		# response = book_pb2.Book(name=None)
		# return response

		super().GetBook(request, context)

File: bookstore-go/services/BookStoreService/v1/BookStoreService.go
package BookStoreService

import (
	"context"

	bookv1 "github.com/sylk-build/sylk-examples/bookstore-go/services/protos/public/bookstore/book/v1"
	"github.com/sylk-build/sylk-examples/bookstore-go/services/utils"
	"google.golang.org/grpc/metadata"
	emptypb "google.golang.org/protobuf/types/known/emptypb"
)

type BookStoreService struct {
	bookv1.UnimplementedBookStoreServiceServer
}

// [sylk.build] - BookStoreService.ListBooks - None
func (BookStoreServiceServicer *BookStoreService) ListBooks(Empty *emptypb.Empty, stream bookv1.BookStoreService_ListBooksServer) (err error) {
	printLog("ListBooks", stream.Context(), nil)
	books := []bookv1.Book{
		{
			Name: "Clean Code: A Handbook of Agile Software Craftsmanship",
		},
		{
			Name: "The Pragmatic Programmer: Your Journey to Mastery",
		},
		{
			Name: "Design Patterns: Elements of Reusable Object-Oriented Software",
		},
	}
	// Iterate over the list of books and send each book as a Book message through the stream
	for _, book := range books {
		if err := stream.Send(&book); err != nil {
			return err
		}
	}

	return nil
}

// [sylk.build] - BookStoreService.GetBook - None
func (BookStoreServiceServicer *BookStoreService) GetBook(ctx context.Context, GetBookRequest *bookv1.GetBookRequest) (response *bookv1.Book, err error) {
	printLog("GetBook", ctx, GetBookRequest)
	return &bookv1.Book{}, nil
}

func printLog(name string, ctx context.Context, message interface{}) {
	contextMetadata, _ := metadata.FromIncomingContext(ctx)
	utils.InfoLogger.Printf("[%s] Got RPC request: %v", name, message)
	utils.DebugLogger.Printf("[%s] Metadata: %v", name, contextMetadata)
}
